### PR TITLE
[SPARK-21024][SQL] CSV parse mode handles Univocity parser exceptions

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr fixed code to handle Univocity parser exceptions by CSV parse modes.
The current master cannot skip the illegal records that Univocity parsers:
```
scala> Seq("0,1", "0,1,2,3").toDF().write.text("/Users/maropu/Desktop/data")
scala> val df = spark.read.format("csv").schema("a int, b int").option("maxColumns", "3").load("/Users/maropu/Desktop/data")
scala> df.show

com.univocity.parsers.common.TextParsingException: java.lang.ArrayIndexOutOfBoundsException - 3
Hint: Number of columns processed may have exceeded limit of 3 columns. Use settings.setMaxColumns(int) to define the maximum number of columns your input can have
Ensure your configuration is correct, with delimiters, quotes and escape sequences that match the input format you are trying to parse
Parser Configuration: CsvParserSettings:
        Auto configuration enabled=true
        Autodetect column delimiter=false
        Autodetect quotes=false
        Column reordering enabled=true
        Empty value=null
        Escape unquoted values=false
        ...

at com.univocity.parsers.common.AbstractParser.handleException(AbstractParser.java:339)
at com.univocity.parsers.common.AbstractParser.handleEOF(AbstractParser.java:195)
at com.univocity.parsers.common.AbstractParser.parseLine(AbstractParser.java:544)
at org.apache.spark.sql.execution.datasources.csv.UnivocityParser.parse(UnivocityParser.scala:191)
at org.apache.spark.sql.execution.datasources.csv.UnivocityParser$$anonfun$5.apply(UnivocityParser.scala:308)
at org.apache.spark.sql.execution.datasources.csv.UnivocityParser$$anonfun$5.apply(UnivocityParser.scala:308)
at org.apache.spark.sql.execution.datasources.FailureSafeParser.parse(FailureSafeParser.scala:60)
at org.apache.spark.sql.execution.datasources.csv.UnivocityParser$$anonfun$parseIterator$1.apply(UnivocityParser.scala:312)
at org.apache.spark.sql.execution.datasources.csv.UnivocityParser$$anonfun$parseIterator$1.apply(UnivocityParser.scala:312)
at scala.collection.Iterator$$anon$12.nextCur(Iterator.scala:434)
at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:440)
at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
...
```

## How was this patch tested?
Added tests in `CSVSuite`.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala [hunk 1]
```diff
@@ -25,6 +25,7 @@ import java.util.Locale
 import scala.util.Try
 import scala.util.control.NonFatal
 
+import com.univocity.parsers.common.TextParsingException
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: It looks this one can be removed.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 2]
```diff
@@ -1174,4 +1174,27 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
...
+          .collect
+      }.getMessage
+      assert(msg.contains("Number of columns processed may have exceeded limit of 3 columns."))
+
+      val df = spark.read.format("csv")
+        .schema(s"a INT, b INT")
+        .option("maxColumns", "2")
+        .option("mode", "PERMISSIVE")
+        .load(path.getAbsolutePath)
+      checkAnswer(df, Row(0, 1) :: Row(null, null) :: Nil)
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: Should we maybe also check what is put in the malformed column?

**Review by HyukjinKwon — COMMENTED** [review]
Sounds okay to me.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala (hunk 1 above)
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: oh, missed. Thanks

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala (hunk 2 above)
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: okay, I'll set `columnNameOfCorruptRecord` column.

**Comment by HyukjinKwon** [comment]
@maropu, just to make sure, do you mind if I ask test this with `wholeFile` too? IIRC, it calls different a code path for parsing tokens.

**Comment by maropu** [comment]
oh, sure. I'll add tests.

**Comment by maropu** [comment]
@HyukjinKwon I tried to fix this case even in `wholeFile` mode though, I easily couldn't do because `CSVParser` in `wholeFile` has state to parse `InputStream` data. Once the parser hits exception, IIUC it cannot restart parsing the input. So, to fix this in `wholeFile`, we might need to refactor the current logic there. IMHO in case that parsing exceptions happen in `whileFile`, we easily couldn't restart to parse because the restart position for parsing seems unclear. WDYT?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 3]
```diff
@@ -1174,4 +1174,33 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
...
+
+        val columnNameOfCorruptRecord = "_unparsed"
+        val df = spark.read.format("csv")
+          .schema(s"a INT, b INT, $columnNameOfCorruptRecord STRING")
+          .option("maxColumns", "2")
+          .option("mode", "PERMISSIVE")
+          .option("columnNameOfCorruptRecord", columnNameOfCorruptRecord)
+          .option("wholeFile", wholeFile)
+          .load(path.getAbsolutePath)
+        checkAnswer(df, Row(0, 1, null) :: Row(null, null, "0,1,2,") :: Nil)
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: @HyukjinKwon weird behaviour..., when we set `maxColumns` in a Univocity parser, it seems `currentParsedContent` returns the (`maxColumns` + 1) elements in inputs.

**Comment by giaosudau** [comment]
I found that if you using inferSchema option it will throw the same error.
Should handle this case too.

**Comment by maropu** [comment]
Yea, I thinks so. But, to fix the `inferSchema` case, we need some more modifications. So, I feel we first will fix this case (schema specified by user), then fix that case in follow-up.

**Comment by maropu** [comment]
@gatorsmile @HyukjinKwon ping

**Comment by HyukjinKwon** [comment]
I am sorry @maropu. I can't think of a good way to handle this for now ... Will be back after thinking more maybe ..

**Comment by maropu** [comment]
ok, thanks!

**Comment by maropu** [comment]
I think we could be back if we have better way to handle this, so I'll close this for now (we better keeping this discussion in jira).

**Pull request closed without merging**